### PR TITLE
box64: Post-0.2.4 refactoring

### PR DESCRIPTION
**Pull request by OPNA2608**
## Description of changes

Various adjustments, following the [recent bump](https://github.com/NixOS/nixpkgs/pull/250877).

- Fetch a patch to fix a regression in SDL1 coverage
- Migrate from `rec` attrset to `finalAttrs` function mkDerivation pattern
- A minor refactor of how the CMake flags are ordered & filled in, when the check phases are run (require built binaries to be executable locally) to fix cross, and an addition of `meta.mainProgram`
- Some additions regarding more exotic platforms
  - RISC-V received dynarec support in 0.2.4, added handling for it & enabled it by default.
    I lack the hardware to test it, but it cross-compiles for me.
  - Added `loongarch64-linux` and `mips64el-linux` to supported platforms, at least on paper. LA64 is explicitly supported with one of the "mega-options", 64-bit little-endian MIPS by virtue of fitting the upstream target description: `non-x86_64 Linux systems, like ARM (host system needs to be 64-bit little-endian)`.
    The cross-compiler toolchain for both of these targets seems to be broken right now so I can't verify if these build.

I'd appreciate *any* suggestions on how to automatically compose `meta.platforms`. The only unwieldy solution could come up with includes duplicating code from `lib/systems/doubles.nix`, due to there being no `64bit` platforms list:
```nix
# Supported on 64-bit, little-endian, Linux hosts
platforms = map lib.systems.parse.doubleFromSystem
  (lib.lists.filter lib.systems.inspect.predicates.is64bit
    (map lib.systems.parse.mkSystemFromString
      (lib.lists.intersectLists lib.platforms.linux lib.platforms.littleEndian)));
```

This *does* give me the correct list of platforms, but feels too ugly to commit. :sweat_smile:

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Result of `nixpkgs-review pr 251662` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>box64</li>
  </ul>
</details>

File: pkgs/applications/emulators/box64/default.nix [hunk 1]
```diff
@@ -1,42 +1,60 @@
...
-    repo = pname;
-    rev = "v${version}";
+    repo = "box64";
+    rev = "v${finalAttrs.version}";
     hash = "sha256-iCZv/WvqZkH6i23fSLA/p0nG5/CgzjyU5glVgje4c3w=";
   };
 
+  patches = [
+    # Fix crash due to regression in SDL1 AudioCallback signature in 0.2.4
+    # Remove when version > 0.2.4
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/box64/default.nix:29]:
```suggestion
    # Remove at 0.2.4 release
```

Comment (OPNA2608) [reply pkgs/applications/emulators/box64/default.nix:29]: It's at 0.2.4 right now.

Comment (AndersonTorres) [reply pkgs/applications/emulators/box64/default.nix:29]: It makes sense now.

**Comment by OPNA2608** [comment]
@matthiasbeyer: If you `nixpkgs-review`ed it with no requests to change anything, can I assume that you meant to approve it?

@AndersonTorres: Is there anything else about the changes that looks off to you?

Thanks.

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
LGTM

**Comment by OPNA2608** [comment]
@gador you co-maintain this package, could you find the time to give this a review so we can get it in for 23.11 please?

**Review by gador — APPROVED** [review]
Sorry for the wait. I also lack the hardware to test the additional supported platforms. FWIW it looks good to me! Tested on `aarch64-linux` (emulated) and `x86-64-linux`.

**Pull request merged**